core/git.go

```diff
@@ -254,92 +463,240 @@ func (ref *RemoteGitRef) PBDefinitions(ctx context.Context) ([]*pb.Definition, e
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	if len(p.Sources.Git) == 0 {
-		return "", errors.Errorf("no git commit was resolved")
+	checkoutRef = nil
+	defer func() {
+		if rerr != nil {
+			snap.Release(context.WithoutCancel(ctx))
+		}
+	}()
+
+	md := cacheRefMetadata{snap}
+	if err := md.setGitSnapshot(cacheKey); err != nil {
+		return nil, err
 	}
-	return p.Sources.Git[0].Commit, nil
+
+	checkout := NewDirectory(ref.Query, nil, "/", ref.Query.Platform(), ref.Repo.Services)
+	checkout.Result = snap
+	return checkout, nil
 }
 
-func (ref *RemoteGitRef) getState(ctx context.Context, discardGitDir bool) (llb.State, error) {
-	opts := []llb.GitOption{}
+func doGitCheckout(
+	ctx context.Context,
+	checkoutGit *gitutil.GitCLI,
+	remote string,
+	fullref string, commit string,
+	depth int,
+	discardGitDir bool,
+) error {
+	checkoutDirGit, err := checkoutGit.GitDir(ctx)
+	if err != nil {
+		return fmt.Errorf("could not find git dir: %w", err)
+	}
 
-	if !discardGitDir {
-		opts = append(opts, llb.KeepGitDir())
+	pullref := fullref
+	if !gitutil.IsCommitSHA(fullref) {
+		pullref += ":" + pullref
 	}
-	if ref.Repo.SSHKnownHosts != "" {
-		opts = append(opts, llb.KnownSSHHosts(ref.Repo.SSHKnownHosts))
+
+	args := []string{"fetch", "-u"}
+	if depth != 0 {
+		args = append(args, fmt.Sprintf("--depth=%d", depth))
+	}
+	args = append(args, "origin", pullref)
+	_, err = checkoutGit.Run(ctx, args...)
+	if err != nil {
+		return err
+	}
+	_, err = checkoutGit.Run(ctx, "checkout", strings.TrimPrefix(fullref, "refs/heads/"))
+	if err != nil {
+		return fmt.Errorf("failed to checkout remote %s: %w", remote, err)
+	}
+	_, err = checkoutGit.Run(ctx, "reset", "--hard", commit)
+	if err != nil {
+		return fmt.Errorf("failed to reset ref %s: %w", remote, err)
 	}
-	if ref.Repo.SSHAuthSocket != nil {
-		opts = append(opts, llb.MountSSHSock(ref.Repo.SSHAuthSocket.LLBID()))
+	_, err = checkoutGit.Run(ctx, "remote", "set-url", "origin", remote)
+	if err != nil {
+		return fmt.Errorf("failed to set remote origin to %s: %w", remote, err)
 	}
-	if ref.Repo.AuthToken.Self != nil {
-		opts = append(opts, llb.AuthTokenSecret(ref.Repo.AuthToken.ID().Digest().String()))
+	_, err = checkoutGit.Run(ctx, "reflog", "expire", "--all", "--expire=now")
+	if err != nil {
+		return fmt.Errorf("failed to expire reflog for remote %s: %w", remote, err)
 	}
-	if ref.Repo.AuthHeader.Self != nil {
-		opts = append(opts, llb.AuthHeaderSecret(ref.Repo.AuthHeader.ID().Digest().String()))
+
+	if err := os.Remove(filepath.Join(checkoutDirGit, "FETCH_HEAD")); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("failed to remove FETCH_HEAD for remote %s: %w", remote, err)
 	}
 
-	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	// TODO: this feels completely out-of-sync from how we do the rest
+	// of the clone - caching will not be as great here
+	_, err = checkoutGit.Run(ctx, "submodule", "update", "--init", "--recursive", "--depth=1")
```

Comment:
Note to the reviewer: the `TODO`s in this PR are mostly intentional, they don't need to be done here, since they're not a regression from the *existing* behavior.

However, just marking them for future improvement.